analyses/cell-type-ewings/scripts/utils/tumor-validation-helpers.R

```diff
@@ -0,0 +1,200 @@
+  plot_markers_df <- classification_df |> 
+    # add in marker gene expression to dataframe 
+    dplyr::left_join(marker_gene_exp, by = "barcodes") |> 
+    # combine all genes into a single column for easy faceting
+    tidyr::pivot_longer(
+      cols = starts_with("ENSG"),
+      names_to = "ensembl_gene_id",
+      values_to = "gene_expression"
+    ) |> 
+    # join with marker gene df to get gene symbols for plotting 
+    dplyr::left_join(marker_genes_df, by = c("ensembl_gene_id")) |> 
+    # keep only gene symbols, barcodes, and classification columns 
+    dplyr::select(barcodes, gene_symbol, gene_expression, ends_with("classification")) |> 
+    unique() |> 
+    dplyr::group_by(gene_symbol) |>
+    # get z-scores for each gene
+    dplyr::mutate(transformed_gene_expression = scale(gene_expression)[, 1]) |> 
+    dplyr::ungroup() |> 
+    dplyr::group_by(barcodes) |> 
+    # add all the scores together 
+    dplyr::mutate(sum_exp = sum(transformed_gene_expression)) |>
+    dplyr::ungroup()
+  
+  return(plot_markers_df)
+}
+
+# Heatmaps ---------------------------------------------------------------------
+
+# create a heatmap where rows are marker genes or gene sets and columns are cells 
+# this function adds a column annotation since cells are columns 
+plot_gene_heatmap <- function(
+    df, 
+    row_title = "",
+    legend_title = "", 
+    annotation = NULL
+){
+  
+  # plot heatmap of marker genes 
+  heatmap <- ComplexHeatmap::Heatmap(
+    df,
+    # set the color scale based on min and max values
+    col = circlize::colorRamp2(seq(min(df), max(df), length = 2), colors = c("white", "#00274C")),
+    border = TRUE,
+    ## Row parameters
+    cluster_rows = TRUE,
+    row_title = row_title,
+    row_title_gp = grid::gpar(fontsize = 12),
+    row_title_side = "left",
+    row_names_side = "left",
+    row_dend_side = "right",
+    row_names_gp = grid::gpar(fontsize = 10),
+    ## Column parameters
+    cluster_columns = TRUE,
+    show_column_names = FALSE,
+    bottom_annotation = annotation,
+    heatmap_legend_param = list(
+      title = legend_title
+    )
+  )
+  
+  return(heatmap)
+}
+
+# create a heatmap where rows are cells and columns are chromosomes 
+# this function adds a row annotation since cells are rows 
+plot_cnv_heatmap <- function(
+    df,
+    cnv_col, 
+    annotation = NULL,
+    legend_title = ""
+){
+  
+  # barcode by chromosome df 
+  cnv_df <- df |> 
+    dplyr::select(barcodes, starts_with(cnv_col)) |> 
+    unique() |> 
+    tibble::column_to_rownames("barcodes")
+  
+  # just get the name of the chromosome as column name 
+  colnames(cnv_df) <- stringr::word(colnames(cnv_df), -1, sep = "_")
+  
+  # do a little bit of column reordering before converting to a matrix
+  cnv_df <- cnv_df |> 
+    dplyr::relocate(chr10, .before = chr11) |> 
+    dplyr::relocate(chr19, .before = chr20) |> 
+    dplyr::relocate(chr21, .before = chr22) |> 
```

Comment:
I'm not sure why these specific columns are mis-ordered, but maybe we should set the full order for generality in case other mis-ordering pops up?
```suggestion
    dplyr::select(paste0("chr", 1:22)) |>
```